src/EVM/Expr.hs

```diff
@@ -777,23 +780,48 @@ getLogicalIdx (GVar _) = internalError "cannot determine addr of a GVar"
 data StorageType = SmallSlot | Array | Map | Mixed | UNK
   deriving (Show, Eq)
 
-safeToDecompose :: Expr a -> Bool
-safeToDecompose inp = result /= Mixed
+-- We can't currently decompose cases when the FULL returned state is equated
+-- This is because the decomposition would need to take into account that ALL
+-- maps/arrays/small-slots need to be equivalent. This could be done, but is left
+-- as a TODO. Currently this only affects equivalence checking as there is no
+-- EVM bytecode to compare the FULL state, so such comparison could only be
+-- generated via hevm itself
+safeToDecomposeProp :: Prop -> Maybe ()
```

Comment:
Is there a reason for returning `Maybe ()` instead of `Bool`?

Reply:
Actually, no! And it's a lot cleaner if it's a Bool. Sorry, fixed :)

I tried doing the same for `safeToDecompose`, but because we need to map it over a Prop, and it takes in an Expr, it'd need one more traversal and I am not sure I should add it... so for the moment, I fixed this one that you highlighted, and left the other one as-is.